File: output/amazon_q/java/medium/q35-find-all-anagrams-in-a-string.java
/** https://leetcode.com/problems/find-all-anagrams-in-a-string */
// Problem: Find All Anagrams in a String

//Problem Description:
//Given two strings s and p, return an array of all the start indices of p's anagrams in s. You may return the answer in any order.

// Constraints:
// 1 <= s.length, p.length <= 3 * 10^4
// s and p consist of lowercase English letters.

//Code Structure:
import java.util.ArrayList;
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.length() == 0 || p == null || p.length() == 0) {
            return result;
        }
        int[] hash = new int[256];
        for (char c : p.toCharArray()) {
            hash[c]++;
        }
        int left = 0, right = 0, count = p.length();
        while (right < s.length()) {
            if (hash[s.charAt(right++)]-- >= 1) {
                count--;
            }
            if (count == 0) {
                result.add(left);
            }
            if (right - left == p.length() && hash[s.charAt(left++)]++ >= 0) {
                count++;
            }
        }
        return result;
    }
}
    

File: output/amazon_q/php/medium/q35-find-all-anagrams-in-a-string.php
<?php
/** https://leetcode.com/problems/find-all-anagrams-in-a-string */
// Problem: Find All Anagrams in a String

//Problem Description:
//Given two strings s and p, return an array of all the start indices of p's anagrams in s. You may return the answer in any order.

// Constraints:
// 1 <= s.length, p.length <= 3 * 10^4
// s and p consist of lowercase English letters.

//Code Structure:
class Solution {

    /**
     * @param String $s
     * @param String $p
     * @return Integer[]
     */
    function findAnagrams($s, $p) {
        $result = [];
        $pLength = strlen($p);
        $sLength = strlen($s);
        $pCount = array_fill(0, 26, 0);
        $sCount = array_fill(0, 26, 0);

        if ($pLength > $sLength) {
            return $result;
        }

        for ($i = 0; $i < $pLength; $i++) {
            $pCount[ord($p[$i]) - ord('a')]++;
            $sCount[ord($s[$i]) - ord('a')]++;
        }

        if ($pCount == $sCount) {
            $result[] = 0;
        }

        for ($i = $pLength; $i < $sLength; $i++) {
            $sCount[ord($s[$i]) - ord('a')]++;
            $sCount[ord($s[$i - $pLength]) - ord('a')]--;

            if ($pCount == $sCount) {
                $result[] = $i - $pLength + 1;
            }
        }

        return $result;
    }
}
    